# Exercise 02 -  OLAP Cubes - Roll Up and Drill Down

All the databases table in this demo are based on public database samples and transformations
- `Sakila` is a sample database created by `MySql` [Link](https://video.udacity-data.com/topher/2021/August/61120e06_pagila-3nf/pagila-3nf.png)
- The postgresql version of it is called `Pagila` [Link](https://github.com/devrimgunduz/pagila)
- The facts and dimension tables design is based on O'Reilly's public dimensional modelling tutorial schema [Link](https://video.udacity-data.com/topher/2021/August/61120e06_pagila-3nf/pagila-3nf.png)

Start by connecting to the database by running the cells below. If you are coming back to this exercise, then uncomment and run the first cell to recreate the database. If you recently completed the slicing and dicing exercise, then skip to the second cell.

In [1]:
!PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila
!PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila -f Data/pagila-schema.sql
!PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila -f Data/pagila-data.sql

 setval 
--------
    200
(1 row)

 setval 
--------
    605
(1 row)

 setval 
--------
     16
(1 row)

 setval 
--------
    600
(1 row)

 setval 
--------
    109
(1 row)

 setval 
--------
    599
(1 row)

 setval 
--------
   1000
(1 row)

 setval 
--------
   4581
(1 row)

 setval 
--------
      6
(1 row)

 setval 
--------
  32098
(1 row)

 setval 
--------
  16049
(1 row)

 setval 
--------
      2
(1 row)

 setval 
--------
      2
(1 row)



### Connect to the local database where Pagila is loaded

In [2]:
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

postgresql://student:student@127.0.0.1:5432/pagila


'Connected: student@pagila'

### Star Schema

<img src="pagila-star.png" width="50%"/>

In [3]:
%%sql
CREATE TABLE dimDate
(
  date_key integer NOT NULL PRIMARY KEY,
  date date NOT NULL,
  year smallint NOT NULL,
  quarter smallint NOT NULL,
  month smallint NOT NULL,
  day smallint NOT NULL,
  week smallint NOT NULL,
  is_weekend boolean
);

CREATE TABLE dimCustomer
(
  customer_key SERIAL PRIMARY KEY,
  customer_id  smallint NOT NULL,
  first_name   varchar(45) NOT NULL,
  last_name    varchar(45) NOT NULL,
  email        varchar(50),
  address      varchar(50) NOT NULL,
  address2     varchar(50),
  district     varchar(20) NOT NULL,
  city         varchar(50) NOT NULL,
  country      varchar(50) NOT NULL,
  postal_code  varchar(10),
  phone        varchar(20) NOT NULL,
  active       smallint NOT NULL,
  create_date  timestamp NOT NULL,
  start_date   date NOT NULL,
  end_date     date NOT NULL
);

CREATE TABLE dimMovie
(
  movie_key          SERIAL PRIMARY KEY,
  film_id            smallint NOT NULL,
  title              varchar(255) NOT NULL,
  description        text,
  release_year       integer,
  language           varchar(20) NOT NULL,
  original_language  varchar(20),
  rental_duration    smallint NOT NULL,
  length             smallint NOT NULL,
  rating             varchar(5) NOT NULL,
  special_features   varchar(60) NOT NULL
);
CREATE TABLE dimStore
(
  store_key           SERIAL PRIMARY KEY,
  store_id            smallint NOT NULL,
  address             varchar(50) NOT NULL,
  address2            varchar(50),
  district            varchar(20) NOT NULL,
  city                varchar(50) NOT NULL,
  country             varchar(50) NOT NULL,
  postal_code         varchar(10),
  manager_first_name  varchar(45) NOT NULL,
  manager_last_name   varchar(45) NOT NULL,
  start_date          date NOT NULL,
  end_date            date NOT NULL
);
CREATE TABLE factSales
(
  sales_key        SERIAL PRIMARY KEY,
  date_key         INT NOT NULL REFERENCES dimDate(date_key),
  customer_key     INT NOT NULL REFERENCES dimCustomer(customer_key),
  movie_key        INT NOT NULL REFERENCES dimMovie(movie_key),
  store_key        INT NOT NULL REFERENCES dimStore(store_key),
  sales_amount     decimal(5,2) NOT NULL
);

 * postgresql://student:***@127.0.0.1:5432/pagila
Done.
Done.
Done.
Done.
Done.


[]

In [4]:
%%sql
INSERT INTO dimDate (date_key, date, year, quarter, month, day, week, is_weekend)
SELECT DISTINCT(TO_CHAR(payment_date :: DATE, 'yyyyMMDD')::integer) AS date_key,
       date(payment_date)                                           AS date,
       EXTRACT(year FROM payment_date)                              AS year,
       EXTRACT(quarter FROM payment_date)                           AS quarter,
       EXTRACT(month FROM payment_date)                             AS month,
       EXTRACT(day FROM payment_date)                               AS day,
       EXTRACT(week FROM payment_date)                              AS week,
       CASE WHEN EXTRACT(ISODOW FROM payment_date) IN (6, 7) THEN true ELSE false END AS is_weekend
FROM payment;


INSERT INTO dimCustomer (customer_key, customer_id, first_name, last_name, email, address, address2, district, city, country, postal_code, phone, active, create_date, start_date, end_date)
SELECT c.customer_id AS customer_key,
       c.customer_id,
       c.first_name,
       c.last_name,
       c.email,
       a.address,
       a.address2,
       a.district,
       ci.city,
       co.country,
       a.postal_code,
       a.phone,
       c.active,
       c.create_date,
       now()         AS start_date,
       now()         AS end_date
FROM customer c
JOIN address a  ON (c.address_id = a.address_id)
JOIN city ci    ON (a.city_id = ci.city_id)
JOIN country co ON (ci.country_id = co.country_id);

INSERT INTO dimMovie (movie_key, film_id, title, description, release_year, language, original_language, rental_duration, length, rating, special_features)
SELECT f.film_id      AS movie_key,
       f.film_id,
       f.title,
       f.description,
       f.release_year,
       l.name         AS language,
       orig_lang.name AS original_language,
       f.rental_duration,
       f.length,
       f.rating,
       f.special_features
FROM film f
JOIN language l              ON (f.language_id=l.language_id)
LEFT JOIN language orig_lang ON (f.original_language_id = orig_lang.language_id);

INSERT INTO dimStore (store_key, store_id, address, address2, district, city, country, postal_code, manager_first_name, manager_last_name, start_date, end_date)
SELECT s.store_id    AS store_key,
       s.store_id,
       a.address,
       a.address2,
       a.district,
       c.city,
       co.country,
       a.postal_code,
       st.first_name AS manager_first_name,
       st.last_name  AS manager_last_name,
       now()         AS start_date,
       now()         AS end_date
FROM store s
JOIN staff st   ON (s.manager_staff_id = st.staff_id)
JOIN address a  ON (s.address_id = a.address_id)
JOIN city c     ON (a.city_id = c.city_id)
JOIN country co ON (c.country_id = co.country_id);

INSERT INTO factSales (date_key, customer_key, movie_key, store_key, sales_amount)
SELECT TO_CHAR(p.payment_date :: DATE, 'yyyyMMDD')::integer AS date_key ,
       p.customer_id                                        AS customer_key,
       i.film_id                                            AS movie_key,
       i.store_id                                           AS store_key,
       p.amount                                             AS sales_amount
FROM payment p
JOIN rental r    ON ( p.rental_id = r.rental_id )
JOIN inventory i ON ( r.inventory_id = i.inventory_id );

 * postgresql://student:***@127.0.0.1:5432/pagila
40 rows affected.
599 rows affected.
1000 rows affected.
2 rows affected.
16049 rows affected.


[]

## Roll-up
- Stepping up the level of aggregation to a large grouping
- e.g.`city` is summed as `country`

TODO: Write a query that calculates revenue (sales_amount) by day, rating, and country. Sort the data by revenue in descending order, and limit the data to the top 20 results. The first few rows of your output should match the table below.

In [5]:
%%time
%%sql

SELECT d.day, m.rating, c.country, SUM(f.sales_amount) as revenue
FROM factSales f
JOIN dimDate d ON f.date_key = d.date_key
JOIN dimCustomer c ON f.customer_key = c.customer_key
JOIN dimMovie m ON f.movie_key = m.movie_key
GROUP BY d.day, m.rating, c.country
ORDER BY revenue DESC
LIMIT 20;

 * postgresql://student:***@127.0.0.1:5432/pagila
20 rows affected.
CPU times: user 2.29 ms, sys: 2.79 ms, total: 5.08 ms
Wall time: 40.4 ms


[(30, 'G', 'China', Decimal('169.67')),
 (30, 'PG', 'India', Decimal('156.67')),
 (30, 'NC-17', 'India', Decimal('153.64')),
 (30, 'PG-13', 'China', Decimal('146.67')),
 (30, 'R', 'China', Decimal('145.66')),
 (30, 'R', 'India', Decimal('143.68')),
 (30, 'G', 'India', Decimal('137.67')),
 (18, 'NC-17', 'India', Decimal('135.75')),
 (30, 'PG', 'China', Decimal('131.72')),
 (21, 'PG-13', 'India', Decimal('128.74')),
 (18, 'PG-13', 'India', Decimal('121.72')),
 (18, 'PG', 'India', Decimal('119.76')),
 (30, 'PG-13', 'India', Decimal('116.72')),
 (21, 'NC-17', 'China', Decimal('115.77')),
 (21, 'R', 'India', Decimal('115.75')),
 (27, 'NC-17', 'India', Decimal('115.72')),
 (17, 'PG-13', 'China', Decimal('111.75')),
 (17, 'NC-17', 'United States', Decimal('109.76')),
 (30, 'NC-17', 'China', Decimal('108.77')),
 (20, 'PG-13', 'India', Decimal('101.79'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>China</td>
        <td>169.67</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG</td>
        <td>India</td>
        <td>156.67</td>
    </tr>
    <tr>
        <td>30</td>
        <td>NC-17</td>
        <td>India</td>
        <td>153.64</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>China</td>
        <td>146.67</td>
    </tr>
    <tr>
        <td>30</td>
        <td>R</td>
        <td>China</td>
        <td>145.66</td>
    </tr>
</tbody></table></div>

## Drill-down
- Breaking up one of the dimensions to a lower level.
- e.g.`city` is broken up into  `districts`

TODO: Write a query that calculates revenue (sales_amount) by day, rating, and district. Sort the data by revenue in descending order, and limit the data to the top 20 results. The first few rows of your output should match the table below.

In [6]:
%%time
%%sql

SELECT d.day, m.rating, c.district, SUM(f.sales_amount) as revenue
FROM factSales f
JOIN dimDate d ON f.date_key = d.date_key
JOIN dimCustomer c ON f.customer_key = c.customer_key
JOIN dimMovie m ON f.movie_key = m.movie_key
GROUP BY d.day, m.rating, c.district
ORDER BY revenue DESC
LIMIT 20;

 * postgresql://student:***@127.0.0.1:5432/pagila
20 rows affected.
CPU times: user 4.75 ms, sys: 0 ns, total: 4.75 ms
Wall time: 61.5 ms


[(30, 'PG-13', 'Southern Tagalog', Decimal('53.88')),
 (30, 'G', 'Inner Mongolia', Decimal('38.93')),
 (30, 'G', 'Shandong', Decimal('36.93')),
 (30, 'NC-17', 'West Bengali', Decimal('36.92')),
 (17, 'PG-13', 'Shandong', Decimal('34.95')),
 (1, 'PG', 'California', Decimal('32.94')),
 (18, 'NC-17', 'So Paulo', Decimal('32.93')),
 (21, 'R', 'So Paulo', Decimal('31.93')),
 (30, 'NC-17', 'Buenos Aires', Decimal('31.93')),
 (30, 'PG', 'Southern Tagalog', Decimal('30.94')),
 (30, 'PG', 'So Paulo', Decimal('30.93')),
 (30, 'R', 'Buenos Aires', Decimal('30.92')),
 (30, 'G', 'California', Decimal('29.95')),
 (18, 'NC-17', 'Maharashtra', Decimal('29.95')),
 (21, 'PG-13', 'Uttar Pradesh', Decimal('29.94')),
 (10, 'R', 'Maharashtra', Decimal('29.93')),
 (20, 'PG', 'Shandong', Decimal('29.93')),
 (18, 'R', 'Sumy', Decimal('28.97')),
 (21, 'PG-13', 'Southern Tagalog', Decimal('28.96')),
 (29, 'NC-17', 'So Paulo', Decimal('28.95'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>district</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Southern Tagalog</td>
        <td>53.88</td>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>Inner Mongolia</td>
        <td>38.93</td>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>Shandong</td>
        <td>36.93</td>
    </tr>
    <tr>
        <td>30</td>
        <td>NC-17</td>
        <td>West Bengali</td>
        <td>36.92</td>
    </tr>
    <tr>
        <td>17</td>
        <td>PG-13</td>
        <td>Shandong</td>
        <td>34.95</td>
    </tr>
</tbody></table></div>